src/uri_utils.cpp

```diff
@@ -0,0 +1,220 @@
+    if (next_sep >= uri.size()) {
+        return decomposed;
+    }
+
+    bool authority_found = false;
+    // Find the authority, which always starts with //
+    // https://datatracker.ietf.org/doc/html/rfc3986#section-3.2
+    if ((i + 1) < uri.size() && uri[i] == '/' && uri[i + 1] == '/') {
+        i += 2; // Past '//'
+        authority_found = true;
+        next_sep = uri.find_first_of("/?#", i);
+    }
+
+    if (authority_found && next_sep != i) {
+        // Non empty authority
+        decomposed.authority.index = i;
+        if (next_sep != npos) {
+            // Discard everything after the authority
+            uri = decomposed.scheme_and_authority = uri.substr(0, next_sep);
+            decomposed.authority.raw = uri.substr(i, next_sep);
+        } else {
+            decomposed.scheme_and_authority = uri;
+            decomposed.authority.raw = uri.substr(i);
+        }
+
+        // Identify userinfo, since the character ':' can correspond to both the
+        // password delimiter or the port delimiter, we must search for @
+        // https://datatracker.ietf.org/doc/html/rfc3986#section-3.2.1
+        auto userinfo_end = uri.find('@', i);
+        if (userinfo_end != npos) {
+            decomposed.authority.userinfo = uri.substr(i, userinfo_end - i);
+            // Find any unexpected characters, technically the ':' is valid and the
+            // password is deprecated so allow one or more instances of it.
+            // ALPHA / DIGIT / "-" / "." / "_" / "~" / "!" / "$" / "&" / "'" /
+            // "(" / ")" / "*" / "+" / "," / ";" / "=" / "%"
+            for (; i < userinfo_end; ++i) {
+                // We've found an invalid character, we can consider the
+                // authority malformed
+                if (!isuserinfochar(uri[i])) {
+                    return std::nullopt;
+                }
+            }
+            i = userinfo_end + 1;
+        }
+
+        // Identify the (non-optional) host
+        // https://datatracker.ietf.org/doc/html/rfc3986#section-3.2.2
+        if (i < uri.size()) {
+            auto host_begin = i;
+            if (uri[i] == '[') {
+                // Validate if this is an IPv6 host
+                bool end_found = false;
+                bool non_ip_chars = false;
+                for (i += 1; i < uri.size(); ++i) {
+                    const auto c = uri[i];
+                    if (c == ']') { /* IPv6 End */
+                        end_found = true;
+                        break;
+                    }
+                    if (!ddwaf::isxdigit(c) && c != ':') {
+                        // The host is already malformed so we can stop here;
+                        non_ip_chars = true;
+                        break;
+                    }
+                }
+                if (!end_found || non_ip_chars || i == (host_begin + 1)) {
+                    return std::nullopt;
+                }
+
+                // Valid IPv6, remove the []
```

Comment:
Fair, we assume it's IPv6 because it has the correct character set and enclosing characters, however the host can't have `[`, `]`, `:`, so I thought it was a fair assumption. We could try to parse it as well if we're being strict.